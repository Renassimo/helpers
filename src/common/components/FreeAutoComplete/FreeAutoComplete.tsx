import { useEffect, useMemo } from 'react';
import Fuse from 'fuse.js';

import Autocomplete from '@mui/material/Autocomplete';
import TextField from '@mui/material/TextField';

import { getMuiColor } from '@/common/utils/mui';

import { Matcher } from '@/common/types/matchers';

export interface FreeAutoCompleteProps {
  setValue: (newValue: string | null) => void;
  options: string[];
  value?: string | null;
  label?: string;
  originalValue?: string | null;
  matchers?: Matcher;
}

const FreeAutoComplete = ({
  setValue,
  options = [],
  value,
  label,
  originalValue,
  matchers,
}: FreeAutoCompleteProps) => {
  const isValueFoundInOptions = !!options.find((option) => option === value);
  const [color, sxColor] = getMuiColor(!!value, isValueFoundInOptions);

  const onChange = (_: unknown, newValue: string | null) => setValue(newValue);

  const helperText = originalValue && value !== originalValue && (
    <span style={{ display: 'block' }} onClick={() => setValue(originalValue)}>
      {originalValue}
    </span>
  );

  const fuse = useMemo(
    () =>
      new Fuse(options as string[], {
        includeScore: true,
      }),
    [options]
  );

  useEffect(() => {
    if (!originalValue) return;
    if (originalValue === value) return;

    if (matchers?.[originalValue]) {
      setValue(matchers[originalValue]);
      return;
    }

    const [found] = fuse.search(originalValue);
    if (!found) return;

    const { score } = found;

    if (typeof score !== 'number') return;
    if (score > 0.52) return;

    setValue(found.item);
  }, [originalValue]);

  return (
    <Autocomplete
      value={value ?? ''}
      options={options}
      onInputChange={onChange}
      onChange={onChange}
      size="small"
      disablePortal
      freeSolo
      blurOnSelect
      clearOnBlur
      sx={{ my: 1, color: sxColor }}
      filterOptions={(options, params) => {
        const { inputValue } = params;

        if (inputValue.length === 0) return options;

        return fuse.search(inputValue).map(({ item }) => item);
      }}
      renderInput={(params) => (
        <TextField
          {...params}
          variant="standard"
          color={color}
          label={label}
          fullWidth
          helperText={helperText}
          FormHelperTextProps={{
            sx: { color: 'warning.dark', cursor: 'pointer' },
          }}
          sx={{
            input: { color: sxColor },
          }}
        />
      )}
    />
  );
};

export default FreeAutoComplete;
