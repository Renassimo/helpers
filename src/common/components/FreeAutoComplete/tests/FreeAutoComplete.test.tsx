import userEvent from '@testing-library/user-event';
import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import FreeAutoComplete, { FreeAutoCompleteProps } from '../FreeAutoComplete';

describe('FreeAutoComplete', () => {
  const label = 'Label';
  const setValue = jest.fn();
  const options = ['Option 1', 'Variant 2', 'Opt 3', 'Var 4'];
  const defaultProps: FreeAutoCompleteProps = {
    setValue,
    options,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders successfully', () => {
    // Arange
    // Act
    const { baseElement } = renderWithTheme(
      <FreeAutoComplete {...defaultProps} />
    );
    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  describe('when clicks to input', () => {
    const props = {
      ...defaultProps,
      label,
    };

    test('renders successfully', async () => {
      // Arange
      const { baseElement, getByLabelText } = renderWithTheme(
        <FreeAutoComplete {...props} />
      );
      // Act
      await userEvent.click(getByLabelText(label));
      // Assert
      expect(baseElement).toMatchSnapshot();
    });

    describe('and filters value', () => {
      test('renders successfully', async () => {
        // Arange
        const { baseElement, getByLabelText } = renderWithTheme(
          <FreeAutoComplete {...props} />
        );
        const input = getByLabelText(label);
        await userEvent.click(input);
        // Act
        await userEvent.type(input, 'Variant');
        // Assert
        expect(baseElement).toMatchSnapshot();
      });

      describe('when chooses option', () => {
        test('calls setValue with passed arguments', async () => {
          // Arange
          const typedText = 'Variant';
          const { getByLabelText, getByText } = renderWithTheme(
            <FreeAutoComplete {...props} />
          );
          const input = getByLabelText(label);
          await userEvent.click(input);
          await userEvent.type(input, typedText);
          // Act
          await userEvent.click(getByText('Var 4'));
          // Assert
          expect(setValue).nthCalledWith(typedText.length + 1, 'Var 4');
        });
      });
    });
  });

  describe('when originalValue passed', () => {
    const originalValue = 'Varianto 2';
    const props = { ...defaultProps, originalValue };

    test('calls setValue with most proper option', () => {
      // Arange
      // Act
      const { baseElement } = renderWithTheme(<FreeAutoComplete {...props} />);
      // Assert
      expect(baseElement).toMatchSnapshot();
      expect(setValue).nthCalledWith(1, 'Variant 2');
    });

    describe('when value and original value are the same', () => {
      test('calls setValue with passed arguments', async () => {
        // Arange
        // Act
        renderWithTheme(<FreeAutoComplete {...props} value={originalValue} />);
        // Assert
        expect(setValue).nthCalledWith(1, originalValue);
      });
    });

    describe('when matchers passed', () => {
      test('calls setValue with passed arguments from matcher', async () => {
        // Arange
        const matchers = {
          'Varianto 2': 'Var 4',
        };
        // Act
        renderWithTheme(<FreeAutoComplete {...props} matchers={matchers} />);
        // Assert
        expect(setValue).nthCalledWith(1, 'Var 4');
      });
    });

    describe('when does not find options', () => {
      test('does not call setValue', async () => {
        // Arange
        // Act
        renderWithTheme(
          <FreeAutoComplete {...props} originalValue={'9999999'} />
        );
        // Assert
        expect(setValue).not.toBeCalled();
      });
    });

    describe('when clicks to helper text', () => {
      test('calls setValue with passed arguments', async () => {
        // Arange
        const { getByText } = renderWithTheme(<FreeAutoComplete {...props} />);
        // Act
        await userEvent.click(getByText(originalValue));
        // Assert
        expect(setValue).nthCalledWith(1, 'Variant 2');
        expect(setValue).nthCalledWith(2, originalValue);
      });
    });
  });
});
