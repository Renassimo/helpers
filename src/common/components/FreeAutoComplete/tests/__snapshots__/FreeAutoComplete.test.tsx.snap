// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreeAutoComplete renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon css-wm3bhv-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-dju713-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nz481w-MuiInputBase-input-MuiInput-input"
            id=":r0:"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Clear"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
              tabindex="-1"
              title="Clear"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`FreeAutoComplete when clicks to input and filters value renders successfully 1`] = `
<body>
  <div>
    <div
      aria-owns=":r4:-listbox"
      class="MuiAutocomplete-root Mui-focused MuiAutocomplete-hasClearIcon css-wm3bhv-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-dju713-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r4:"
          id=":r4:-label"
        >
          Label
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-controls=":r4:-listbox"
            aria-expanded="true"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nz481w-MuiInputBase-input-MuiInput-input"
            id=":r4:"
            role="combobox"
            spellcheck="false"
            type="text"
            value="Variant"
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Clear"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
              tabindex="-1"
              title="Clear"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiAutocomplete-popper MuiAutocomplete-popperDisablePortal css-1rndd9j-MuiPopper-root-MuiAutocomplete-popper MuiPopperUnstyled-root"
      data-popper-escaped=""
      data-popper-placement="bottom"
      data-popper-reference-hidden=""
      role="presentation"
      style="position: absolute; top: 0px; left: 0px; width: 0px; margin: 0px; transform: translate(0px, 0px);"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAutocomplete-paper css-9e5uuu-MuiPaper-root-MuiAutocomplete-paper"
      >
        <ul
          aria-labelledby=":r4:-label"
          class="MuiAutocomplete-listbox css-gdh49b-MuiAutocomplete-listbox"
          id=":r4:-listbox"
          role="listbox"
        >
          <li
            aria-disabled="false"
            aria-selected="false"
            class="MuiAutocomplete-option"
            data-option-index="0"
            id=":r4:-option-0"
            role="option"
            tabindex="-1"
          >
            Variant 2
          </li>
          <li
            aria-disabled="false"
            aria-selected="false"
            class="MuiAutocomplete-option"
            data-option-index="1"
            id=":r4:-option-1"
            role="option"
            tabindex="-1"
          >
            Var 4
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`FreeAutoComplete when clicks to input renders successfully 1`] = `
<body>
  <div>
    <div
      aria-owns=":r2:-listbox"
      class="MuiAutocomplete-root Mui-focused MuiAutocomplete-hasClearIcon css-wm3bhv-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-dju713-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r2:"
          id=":r2:-label"
        >
          Label
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-controls=":r2:-listbox"
            aria-expanded="true"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nz481w-MuiInputBase-input-MuiInput-input"
            id=":r2:"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Clear"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
              tabindex="-1"
              title="Clear"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiAutocomplete-popper MuiAutocomplete-popperDisablePortal css-1rndd9j-MuiPopper-root-MuiAutocomplete-popper MuiPopperUnstyled-root"
      data-popper-escaped=""
      data-popper-placement="bottom"
      data-popper-reference-hidden=""
      role="presentation"
      style="position: absolute; top: 0px; left: 0px; width: 0px; margin: 0px; transform: translate(0px, 0px);"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAutocomplete-paper css-9e5uuu-MuiPaper-root-MuiAutocomplete-paper"
      >
        <ul
          aria-labelledby=":r2:-label"
          class="MuiAutocomplete-listbox css-gdh49b-MuiAutocomplete-listbox"
          id=":r2:-listbox"
          role="listbox"
        >
          <li
            aria-disabled="false"
            aria-selected="false"
            class="MuiAutocomplete-option"
            data-option-index="0"
            id=":r2:-option-0"
            role="option"
            tabindex="-1"
          >
            Option 1
          </li>
          <li
            aria-disabled="false"
            aria-selected="false"
            class="MuiAutocomplete-option"
            data-option-index="1"
            id=":r2:-option-1"
            role="option"
            tabindex="-1"
          >
            Variant 2
          </li>
          <li
            aria-disabled="false"
            aria-selected="false"
            class="MuiAutocomplete-option"
            data-option-index="2"
            id=":r2:-option-2"
            role="option"
            tabindex="-1"
          >
            Opt 3
          </li>
          <li
            aria-disabled="false"
            aria-selected="false"
            class="MuiAutocomplete-option"
            data-option-index="3"
            id=":r2:-option-3"
            role="option"
            tabindex="-1"
          >
            Var 4
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`FreeAutoComplete when originalValue passed calls setValue with most proper option 1`] = `
<body>
  <div>
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon css-wm3bhv-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-dju713-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-describedby=":r8:-helper-text"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nz481w-MuiInputBase-input-MuiInput-input"
            id=":r8:"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Clear"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
              tabindex="-1"
              title="Clear"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeSmall css-1a49wnf-MuiFormHelperText-root"
          id=":r8:-helper-text"
        >
          <span
            style="display: block;"
          >
            Varianto 2
          </span>
        </p>
      </div>
    </div>
  </div>
</body>
`;
